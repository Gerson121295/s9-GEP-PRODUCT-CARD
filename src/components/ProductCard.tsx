
import styles from '../styles/styles.module.css'
import { useProduct } from '../hooks/useProduct';
import React, { createContext, JSX} from 'react';
import { InitialValues, onChangeArgs, Product, ProductCardHandlers, ProductContextProps } from '../interfaces/interfaces';


//Context para compartir informacion en los componentes hijos
export const ProductContext = createContext({} as ProductContextProps);

const { Provider} = ProductContext; //Provider es el proveedor del ProductContext


//Definicion de propiedades que recibe el componente ProductCard
export interface Props {
    product: Product;
    //children?: ReactElement | ReactElement[]; //viene un hijo o un arreglo de hijos
    children: ( args: ProductCardHandlers ) => JSX.Element, //children es una funcion que retorna un JSX.Element eb ShoppingPage
    className?: string;  //forma 1 de agregar style mediante className
    style?:  React.CSSProperties; //forma 2 de agregar style mediante style. para saber que style es tipo React.CSSProperties | undefined en un div se agrega la etiqueta style y luego se pasa el cursos y aparece el tipo de dato
    onChange?: (args: onChangeArgs) => void;  //se creo una interfaz con los args. otra forma es definirlo aqui: (product:Product, count :number) => void; //es una funcion que  recibe 2 parametros y no retorna nada
    value?: number; //cantidad del producto en el carrito de compras
    initialValues?: InitialValues; //initialValues definida en la interfaz InitialValues 
  }


export const ProductCard = ({ children, product, className, style, onChange, value, initialValues }: Props) => {

    //customHook useProduct
    const {counter, increaseBy, maxCount, isMaxCountReached, reset, 
    }  = useProduct({onChange, product, value, initialValues }); //se pasa el valor de onChange, product y value al hook useProduct
    
  return (

    <Provider value={{ //value es la informacion que el Provider comparte a los hijos de ProductCard
        counter,
        product, 
        maxCount,
        increaseBy, 
    }}> 
        <div 
          className={`${ styles.productCard } ${ className }`} //por className se pasa el style definido en el padre ShoppingPage.tsx
          style={style} //se agrega el style proveniente de la prop style definido en el padre ShoppingPage.tsx
        >
            
            {
              children({ //se ejecuta la funcion children que retorna un JSX.Element en ShoppingPage.tsx
                count: counter,
                isMaxCountReached,
                maxCount: initialValues?.maxCount,
                product,

                increaseBy,
                reset: reset,
            })
          }

        </div>
    </Provider>
  )
}


