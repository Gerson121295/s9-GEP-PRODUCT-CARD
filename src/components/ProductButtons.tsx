
import React, { useCallback, useContext } from 'react';
import styles from '../styles/styles.module.css'
import { ProductContext } from './ProductCard';

export interface Props{
    className? :string; //forma 1 de agregar style mediante className
    style?:  React.CSSProperties; //forma 2 de agregar style mediante style. para saber que style es tipo React.CSSProperties | undefined en un div se agrega la etiqueta style y luego se pasa el cursos y aparece el tipo de dato
}

export const ProductButtons = ( {className, style} :Props ) => {    //  ({counter, increaseBy}: ProductButtonsProps) => {

    //useContext para extraer datos: counter, increaseBy evitando asi recibirlo por argumento
    const { increaseBy, counter, maxCount } = useContext(ProductContext);

    //useCallback para memorizar una funcion y no se vuelva a renderizar
    const isMaxReached = useCallback(
      () => !!maxCount && counter === maxCount, //si maxCount existe (!!maxCount) entonces evalua si counter es igual a maxCount, si es regresa true
      [counter, maxCount], //memoria la funcion cuando counter y maxCount cambian
    )
    
    return(
        <div 
            className={`${styles.buttonsContainer} ${className}`}
            style={style} //se agrega el style proveniente de la prop style definido en el padre ShoppingPage.tsx
        >
        <button 
            className={styles.buttonMinus}
            onClick={ () => increaseBy( -1 )}
        > - </button>

        <div className={styles.countLabel}> {counter} </div>

        <button 
            className={ `${ styles.buttonAdd }  ${ isMaxReached() && styles.disabled}`} //si isMaxReached es true se agrega la clase disabled
            onClick={ () => increaseBy (+1)}
        > + </button>
    </div>
    )
}

