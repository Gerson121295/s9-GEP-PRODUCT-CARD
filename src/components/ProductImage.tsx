
import React, { useContext } from "react";
import { ProductContext } from "./ProductCard"; 
import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';

export interface Props {
    className?: string;
    img?: string;
    style?:  React.CSSProperties; //forma 2 de agregar style mediante style. para saber que style es tipo React.CSSProperties | undefined en un div se agrega la etiqueta style y luego se pasa el cursos y aparece el tipo de dato
}


export const ProductImage = ({ img, className, style } :Props) => { //{ img = '' }  //{product}: Props  

    const {product} = useContext(ProductContext)

    let imgToShow: string;

    if (img){
        imgToShow = img; //si imagen es true o vine por props del padre entonces imgToShow es igual a img
    }else if (product.img){
        imgToShow = product.img; //si product.img es true o viene del context entonces imgToShow es = a product.img
    }else{ 
        imgToShow = noImage;  //en caso contrario de no venir la imagen imgToShow es = a noImage
    }

    return (
        <img 
            //className={styles.productImg}  //sin el className enviado desde el padre ShoppingPage
            className={`${styles.productImg} ${className}`}  //esta forma permite definir estilos desde ShoppingPage a ProductImage
            style={style} //se agrega el style proveniente de la prop style definido en el padre ShoppingPage.tsx
            src={ imgToShow } 
            alt="Product" 
        /> 
    )
}