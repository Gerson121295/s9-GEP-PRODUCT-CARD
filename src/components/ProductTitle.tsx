import React, { useContext } from "react";
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css'

export interface Props{
    className?: string,
    title?: string,
    activeClass?: string; //ejemplo de que aqui se puede agregar mas elementos ya que esta interfaz es definida en interfaces.ts
    style?:  React.CSSProperties; //forma 2 de agregar style mediante style. para saber que style es tipo React.CSSProperties | undefined en un div se agrega la etiqueta style y luego se pasa el cursos y aparece el tipo de dato
}


export const ProductTitle = ({title, className, style} :Props) => {        //({title}: {title?:string}) => {  //cuando es de 2 o mas ya se crea una interface

    const {product} = useContext(ProductContext);

    return(
        <span 
            className={`${styles.productDescription} ${className}`}
            style={style} //se agrega el style proveniente de la prop style definido en el padre ShoppingPage.tsx
        > 
            { 
                title ? title : product.title //si title viene por props del padre lo usa sino utiliza el titulo del product.title
            } 
        </span>
    );
}