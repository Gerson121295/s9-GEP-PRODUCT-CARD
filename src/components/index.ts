
//archivo index.ts de la carpeta components permite exportar todos los archivos de esa carpeta

//export { ProductCard } from "./ProductCard";

import { ProductCard as ProductCardHOC } from './ProductCard';

import { ProductButtons } from './ProductButtons';
import { ProductImage } from './ProductImage';
import { ProductTitle } from './ProductTitle';
import { ProductCardHOCProps } from '../interfaces/interfaces';

export { ProductButtons } from "./ProductButtons";
export { ProductImage } from "./ProductImage";
export { ProductTitle } from "./ProductTitle";


//Config para que en ShoppingPage se pueda agregar a ProducCard de esta forma <ProductCard.Image /> 
//para esta conf es necesario los import 

//en javascript todos los datos excepto por los primitivos son objetos
export const ProductCard: ProductCardHOCProps = Object.assign(ProductCardHOC, {
    Title: ProductTitle,
    Image: ProductImage,
    Buttons: ProductButtons,
})


export default ProductCard;
