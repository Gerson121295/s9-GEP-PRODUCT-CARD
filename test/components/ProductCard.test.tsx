import React from "react";
import { fireEvent, render } from "@testing-library/react";
import { ProductCard  } from "../../src/components";
import { product1 } from "../data/products";


describe('ProductImage', () => {

    test('Debe mostrar el componente correctamente', () => {
        
        //se desestructura de render el container
        const { asFragment } = render(
        <ProductCard product={ product1 }>
            {
                () => (
                    <h1>Product Card</h1>
                )
            }
        </ProductCard>
        );
        expect(asFragment()).toMatchSnapshot();  
    });


  test('Debe de incrementar el contador', () => {
    const { asFragment, container } = render(
        <ProductCard product={ product1 }>
            {
                ({count, increaseBy}) => (
                    <>
                        <h1>Product Card</h1>
                        <span>{count}</span>
                        <button onClick={ () => increaseBy(1)} ></button>
                    </>
                )
            }
        </ProductCard>
        );

        expect(asFragment()).toMatchSnapshot();
 
        const initialCount = container.querySelector('span');
        expect(initialCount?.textContent).toBe('0'); //asigna el span a 0
     
        console.log(container.innerHTML); //imprime el html del container span para ver el valor

        //fireEvent simula un click en el boton aumenta a 1
        fireEvent.click(container.querySelector('button')!);
     
        const updatedCount = container.querySelector('span');
        console.log(container.innerHTML);

        //valida que el span tenga el valor 1
        expect(updatedCount?.textContent).toBe('1'); 

    });

});