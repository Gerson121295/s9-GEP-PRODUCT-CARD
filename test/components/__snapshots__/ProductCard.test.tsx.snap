// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage Debe de incrementar el contador 1`] = `
<DocumentFragment>
  <div
    class="productCard undefined"
  >
    <h1>
      Product Card
    </h1>
    <span>
      0
    </span>
    <button />
  </div>
</DocumentFragment>
`;

exports[`ProductImage Debe mostrar el componente correctamente 1`] = `
<DocumentFragment>
  <div
    class="productCard undefined"
  >
    <h1>
      Product Card
    </h1>
  </div>
</DocumentFragment>
`;
